walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,127 @@
+
+    ArrayList<TimeRange> availableTimeRanges = getAvailableTimeRanges(events, request, request.getDuration());
+    return availableTimeRanges;
+
+  }
+
+  // returns what events have attendees going to the requested event
+  public ArrayList<Event> checkMatchingAttendees(Collection<Event> events, MeetingRequest request) {
+    Collection<String> requestAttendees = request.getAttendees();
+    ArrayList<Event> eventsWithMatchingAttendees = new ArrayList<Event>();
+    
+    for (String attendee : requestAttendees) {
+      for (Event event : events) {
+        ArrayList<String> eventAttendees = new ArrayList<String>(event.getAttendees());
+        if (eventAttendees.contains(attendee)) {
+          eventsWithMatchingAttendees.add(event);
+        }
+      }
+    }
+
+    return eventsWithMatchingAttendees;
+  }
+
+//   public ArrayList<TimeRange> getincovenientTimeRangesSorted(Collection<Event> events) {
+//     ArrayList<TimeRange> incovenientTimeRanges = new ArrayList<TimeRange>();
+
+//     for (Event event : events) {
+//       incovenientTimeRanges.add(event.getWhen());    
+//     }
+
+//     Collections.sort(incovenientTimeRanges, TimeRange.ORDER_BY_START);
+//     return incovenientTimeRanges;
+//   }
+
+  private ArrayList<TimeRange> getIncovenientTimeRanges(Collection<Event> events, Collection<String> requestAttendees) {
+    HashSet<String> requestAttendeesSet = new HashSet<String>();
+
+    for (String attendee : requestAttendees) {
+      requestAttendeesSet.add(attendee);
+    }
+
+    ArrayList<TimeRange> incovenientTimeRanges = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      TimeRange eventTimeRange = event.getWhen();
+      for (String attendee : eventAttendees) {
+        if (eventAttendees.contains(attendee)) {
+          incovenientTimeRanges.add(eventTimeRange);
+        }      
+      }
+    }
+
+    Collections.sort(incovenientTimeRanges, TimeRange.ORDER_BY_START);
+    return incovenientTimeRanges;
   }
+
+
+  public ArrayList<TimeRange> getAvailableTimeRanges(Collection<Event> events, MeetingRequest request, long duration) {
+    
+    ArrayList<TimeRange> availableTimeRanges = new ArrayList<TimeRange>();
+    // possible timespot meetings can start
+    int startOfPossibleTimes = TimeRange.START_OF_DAY;
+    int endOfPossibleTimes = TimeRange.END_OF_DAY;
+    Collection<String> attendees = request.getAttendees();
+    ArrayList<Event> eventsWithMatchingAttendees = checkMatchingAttendees(events, request);
+
+
+    // test cases check
+    if (duration > FULL_DAY) {
+      return availableTimeRanges;
+    }
+    
+    if (eventsWithMatchingAttendees.isEmpty()){
+      availableTimeRanges.add(TimeRange.WHOLE_DAY);
+      return availableTimeRanges;
+    }
+
+    if (attendees.isEmpty()) {
+      availableTimeRanges.add(TimeRange.WHOLE_DAY);
+      return availableTimeRanges;
+    }
+
+    if (events.isEmpty()) {
```

Comment:
This, and the previous three if's can probably be combined with "or" conditionals (||).

```
if (eventsWithMatchingAttendees.isEmpty() || attendees.isEmpty() || events.isEmpty()) {
  availableTimeRanges.add(TimeRange.WHOLE_DAY);
  return availabileTimeRanges;
}
```